ratelimits/bucket.go

```diff
@@ -1,52 +1,253 @@
-		limit:     NewRegistrationsPerIPv6Range,
-		bucketKey: joinWithColon(NewRegistrationsPerIPv6Range.EnumString(), id),
+	return bucketId{
+		limitName: name,
+		bucketKey: joinWithColon(name.EnumString(), id),
 	}, nil
 }
+
+// Transaction is used to represent a single rate limit Transaction. It contains
+// the bucketId and cost of the Transaction, and flags indicating how the
+// Transaction should be processed. Acceptable check and spend combinations are:
+//   - check-and-spend: when check and spend are true, the cost will be checked
+//     against the bucket's capacity and spent/refunded, when possible.
+//   - check-only: when only check is true, the cost will be checked against the
+//     bucket's capacity, but will never be spent/refunded.
+//   - spend-only: when only spend is true, spending is best-effort. Regardless
+//     of the bucket's capacity, the transaction will be considered "allowed".
+type Transaction struct {
+	bucketId
+	cost  int64
+	check bool
+	spend bool
+}
+
+func (t Transaction) checkOnly() bool {
+	return t.check && !t.spend
+}
+
+func (t Transaction) spendOnly() bool {
+	return t.spend && !t.check
+}
+
+func newTransaction(b bucketId, cost int64) Transaction {
+	return Transaction{
+		bucketId: b,
+		cost:     cost,
+		check:    true,
+		spend:    true,
+	}
+}
+
+func newCheckOnlyTransaction(b bucketId, cost int64) Transaction {
+	return Transaction{
+		bucketId: b,
+		cost:     cost,
+		check:    true,
+	}
+}
+
+func newSpendOnlyTransaction(b bucketId, cost int64) Transaction {
+	return Transaction{
+		bucketId: b,
+		cost:     cost,
+		spend:    true,
+	}
+}
+
+// RegistrationsPerIPAddressTransaction returns a Transaction for the
+// NewRegistrationsPerIPAddress limit for the provided IP address.
+func RegistrationsPerIPAddressTransaction(ip net.IP, cost int64) (Transaction, error) {
+	bucketId, err := newIPAddressBucketId(NewRegistrationsPerIPAddress, ip)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newTransaction(bucketId, cost), nil
+}
+
+// RegistrationsPerIPv6RangeTransaction returns a Transaction for the
+// NewRegistrationsPerIPv6Range limit for the /48 IPv6 range which contains the
+// provided IPv6 address.
+func RegistrationsPerIPv6RangeTransaction(ip net.IP, cost int64) (Transaction, error) {
+	bucketId, err := newIPv6RangeCIDRBucketId(NewRegistrationsPerIPv6Range, ip)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newTransaction(bucketId, cost), nil
+}
+
+// OrdersPerAccountTransaction returns a Transaction for the NewOrdersPerAccount
+// limit for the provided ACME registration Id.
+func OrdersPerAccountTransaction(regId, cost int64) (Transaction, error) {
+	bucketId, err := newRegIdBucketId(NewOrdersPerAccount, regId)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newTransaction(bucketId, cost), nil
+}
+
+// FailedAuthorizationsPerAccountCheckOnlyTransaction returns a 0 cost
+// check-only Transaction for the provided ACME registration Id for the
```

Comment:
What are the semantics of checking a 0 cost Transaction? Doesn't that mean "check if there is room to spend 0 in the bucket," which would always evaluate to "yes there's room, go ahead?"